addons/account/models/account_journal.py

```diff
@@ -419,7 +419,7 @@ def write(self, vals):
                     bank_account = self.env['res.partner.bank'].browse(vals['bank_account_id'])
                     if bank_account.partner_id != company.partner_id:
                         raise UserError(_("The partners of the journal's company and the related bank account mismatch."))
-            if 'alias_name' in vals:
+            if 'alias_name' in vals or ('type' in vals and vals['type'] in ('purchase', 'sale')):
```

Comment:
If type is not provided in the vals, you should check the value actually on the record

Reply:
Yes, indeed, I also check that `'alias_id'` is not in vals because In the create function, the `" account.journal "` will first be created and then its `" alias_id "` will be created and set, so we will go into the write function with `vals {'alias_id': id}`, If the `account.journal` has a type = purchase or sale, the condition will be true, so the `"_update_mail_alias"` function will be called again and create a new`" mail_alias "` with the same name because the `self.alias_id` condition will be false:
https://github.com/odoo/odoo/blob/14.0/addons/account/models/account_journal.py#L390-L395